Fix macOS SCEP client issues

hslatman opened the pull request:
This PR adds the intermediate CA certificate to ClientCAs for chain building and sends root certificates to clients when configured in the SCEP provisioner.

It also includes an additional knob to control the encryption algorithm for clients that don't support the default (DES-CBC). 

The SCEP provisioner has now been successfully tested with micromdm/scep, certnanny/sscep and the macOS SCEP client. The macOS SCEP client has also been tested with HTTPS requests instead of just HTTP successfully.

Fixes #746

maraino commented on authority/authority.go:
```diff
@@ -601,6 +595,12 @@ func (a *Authority) IsRevoked(sn string) (bool, error) {
 
+// GetIntermediateCertificate returns the x509 intermediate CA
+// certificate.
+func (a *Authority) GetIntermediateCertificate() (*x509.Certificate, error) {
+	return pemutil.ReadCertificate(a.config.IntermediateCert)
+}
```
Not all the configurations of step-ca will have an intermediate file. For example, by default, an RA will not, and SCEP is a great candidate for an RA, although it might not be supported yet because it requires an RSA certificate and key for decrypt/encrypt things.

To be future proof, we should get the intermediate from `authority.x509CAService`, but right now it does not have a way to do it directly. The other option is to get it from the step-ca own certificate, generated in `authority.GetTLSCertificate()`.

hslatman replied:
There are indeed improvements to how SCEP can be configured and used. I think that should be combined in a bigger package of work related to SCEP after discussing what types of use cases to support and how.

maraino commented on ca/ca.go:
```diff
@@ -432,6 +427,19 @@ func (ca *CA) getTLSConfig(auth *authority.Authority) (*tls.Config, error) {
+
+	// adding the intermediate CA to the pool will allow clients that do not
+	// send the intermediate for chain building to connect to the CA successfully.
+	cert, err := auth.GetIntermediateCertificate()
+	if err != nil {
+		return nil, err
+	}
+	certPool.AddCert(cert)
```
As an easy fix, without requiring that new method or changes in the CAS services, we can get the cert from the tlsCert from above.

maraino replied:
We should also include all intermediates, as there might be more than one.

hslatman replied:
Fixed in https://github.com/smallstep/certificates/pull/797/commits/64680bb16dd076a33e3c1ad53b71b507825c6560.

maraino commented on authority/provisioner/scep.go:
```diff
@@ -18,13 +18,20 @@ type SCEP struct {
 	ChallengePassword string   `json:"challenge,omitempty"`
 	Capabilities      []string `json:"capabilities,omitempty"`
+	// IncludeRoots makes the provisioner return the CA root(s) in the GetCACerts response
+	IncludeRoots bool `json:"includeRoots,omitempty"`
```
I believe this should be called `IncludeRoot`, the ca.json supports multiple roots for rotation purposes, but only one is really in use.

maraino commented on scep/authority.go:
```diff
@@ -153,14 +151,28 @@ func (a *Authority) GetCACertificates() ([]*x509.Certificate, error) {
+	// Clients are responsible to select the right cert(s) to use, though.
+	if p.ShouldIncludeRootsInChain() && len(a.caCerts) > 1 {
+		certs = append(certs, a.caCerts[1:]...)
+	}
```
Perhaps we should only add the actual root and not all of them.

hslatman replied:
Changed in https://github.com/smallstep/certificates/pull/797/commits/64680bb16dd076a33e3c1ad53b71b507825c6560. 

I still have my doubts if what we're doing here is in fact what is expected in SCEP, as what they refer to in the RFC as CA seems to be our intermediate CA, but this change shouldn't break things for people already using SCEP.

maraino replied:
This works in most cases, but it's not strictly right that the first root in `authority.rootX509Certs` is the root that signed the intermediate. But as this special case must be configured in the provisioner, we can force the user to configure the roots accordingly.

maraino reviewed (changes requested):
Although it works for SCEP that's not the proper way to get the intermediate certificate. We don't really have one proper way, but for this case, we can also hack it and get it from the server cert.

maraino reviewed (changes requested):
All ok, but as we have included changes in the provisioner we need to do a couple of changes more:
* Add field to linkedca https://github.com/smallstep/linkedca/blob/be128d16de4d98ef02c31f6deef571997a781e1b/provisioners.proto#L138-L143
* Add changes in ProvisionerToCertificates and ProvisionerToLinkedca in https://github.com/smallstep/certificates/blob/master/authority/provisioners.go

hslatman commented:
@maraino: added the additional fields and updated the implementation. Also opened a PR  for SCEP provisioner CRUD support in the CLI: https://github.com/smallstep/cli/pull/616. 

In the process I had to update the `grpc.WithInsecure` option too, which was deprecated recently.

maraino reviewed (approved):
LGTM: are you still using `LoadProvisionerByID`?

hslatman commented:
@maraino: no, I changed it to the name variant to make it work with LinkedCA. Should I use the ID one instead? Or are you asking because if I don't use it, I should remove it completely?
